vsphere/resource_vsphere_virtual_machine.go

```diff
@@ -756,6 +791,156 @@ func resourceVSphereVirtualMachineUpdate(d *schema.ResourceData, meta interface{
 		})
 	}
 
+	// USB controller devices. Check for changes and apply them.
+	if d.HasChange("usb_controller") {
+		old, new := d.GetChange("usb_controller")
+		oldUSB := old.([]interface{})
+		newUSB := new.([]interface{})
+
+		// Initialize a key counter.
+		keyCounter := -100
+
+		// Initialize controller presence flags.
+		usb2ControllerPresent := false
+		usb3ControllerPresent := false
+
+		// Map to store existing USB controllers and their keys.
+		existingUSBControllers := make(map[string]int32)
+
+		// Check for existing USB controllers and store their keys.
+		for _, dev := range vprops.Config.Hardware.Device {
+			switch controller := dev.(type) {
+			case *types.VirtualUSBController:
+				usb2ControllerPresent = true
+				existingUSBControllers["2.0"] = controller.Key
+			case *types.VirtualUSBXHCIController:
+				usb3ControllerPresent = true
+				existingUSBControllers["3.x"] = controller.Key
+			}
+		}
+
+		// Remove USB controllers that are no longer in the configuration.
+		for _, oldUSBControllerInterface := range oldUSB {
+			oldUSBController := oldUSBControllerInterface.(map[string]interface{})
+			oldUSBVersion := oldUSBController["version"].(string)
+
+			found := false
+			for _, newUSBControllerInterface := range newUSB {
+				newUSBController := newUSBControllerInterface.(map[string]interface{})
+				newUSBVersion := newUSBController["version"].(string)
+				if oldUSBVersion == newUSBVersion {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				var device types.BaseVirtualDevice
+				switch oldUSBVersion {
+				case "2.0":
+					if key, ok := existingUSBControllers["2.0"]; ok {
+						device = &types.VirtualUSBController{
+							VirtualController: types.VirtualController{
+								VirtualDevice: types.VirtualDevice{
+									Key: key,
+								},
+							},
+						}
+					}
+				case "3.1", "3.2":
+					if key, ok := existingUSBControllers["3.x"]; ok {
+						device = &types.VirtualUSBXHCIController{
+							VirtualController: types.VirtualController{
+								VirtualDevice: types.VirtualDevice{
+									Key: key,
+								},
+							},
+						}
+					}
+				default:
+					return fmt.Errorf("unsupported USB version: %s", oldUSBVersion)
+				}
+
+				if device != nil {
+					// Power off the virtual machine before removing the USB controller.
```

Comment:
Don't do this. The machine gets reconfigured at the end of the function.

You are appending device changes but you are not actually reconfiguring the machine in your diff, doing a power cycle achieves nothing

Reply:
I tested without and it does not power off with be existing code. So not sure what is broke but this does not work without the code being added.

Reply:
I think that this might be needed for now @spacegospod due to some existing issues in the provider. I would recommend leaving this in (for now) with a `// TODO` to revisit the concern and reference a tracking PR to review the reasons power on/off might not work currently as expected.

Reply:
> I tested without and it does not power off with the existing code. So not sure what is broke but this does not work without the code being added.

This is a snippet with the conditions for the existing power off routine
```
	if changed || len(spec.DeviceChange) > 0 {
		// Check to see if we need to shutdown the VM for this process.
		if d.Get("reboot_required").(bool) && vprops.Runtime.PowerState != types.VirtualMachinePowerStatePoweredOff {
```
I'm sure your code passes the first statement since you're adding to `spec.DeviceChange`,
The likely culprit is `reboot_required` not being set. If a reboot of the VM is indeed necessary for a USB controller change you should probably mark the new settings as "required with" reboot_required. This way a user would have to make a 
 conscious decision to do a power cycle of their workload.

Reply:
@burnsjared0415 give this a shot:

```go
	if d.HasChange("usb_controller") {
		old, new := d.GetChange("usb_controller")
		oldUSB := old.([]interface{})
		newUSB := new.([]interface{})

		keyCounter := -100
		usb2ControllerPresent := false
		usb3ControllerPresent := false

		// Map to store existing USB controllers and their keys.
		existingUSBControllers := make(map[string]int32)

		// Check for existing USB controllers and store their keys.
		for _, dev := range vprops.Config.Hardware.Device {
			switch controller := dev.(type) {
			case *types.VirtualUSBController:
				usb2ControllerPresent = true
				existingUSBControllers["2.0"] = controller.Key
			case *types.VirtualUSBXHCIController:
				usb3ControllerPresent = true
				existingUSBControllers["3.x"] = controller.Key
			}
		}

		rebootRequired := false

		// Remove USB controllers that are no longer in the configuration.
		for _, oldUSBControllerInterface := range oldUSB {
			oldUSBController := oldUSBControllerInterface.(map[string]interface{})
			oldUSBVersion := oldUSBController["version"].(string)

			found := false
			for _, newUSBControllerInterface := range newUSB {
				newUSBController := newUSBControllerInterface.(map[string]interface{})
				newUSBVersion := newUSBController["version"].(string)
				if oldUSBVersion == newUSBVersion {
					found = true
					break
				}
			}

			if !found {
				key, exists := existingUSBControllers[oldUSBVersion]
				if exists {
					var device types.BaseVirtualDevice

					switch oldUSBVersion {
					case "2.0":
						device = &types.VirtualUSBController{
							VirtualController: types.VirtualController{
								VirtualDevice: types.VirtualDevice{
									Key: key,
								},
							},
						}
					case "3.1", "3.2":
						device = &types.VirtualUSBXHCIController{
							VirtualController: types.VirtualController{
								VirtualDevice: types.VirtualDevice{
									Key: key,
								},
							},
						}
					}

					if device != nil {
						spec.DeviceChange = append(spec.DeviceChange, &types.VirtualDeviceConfigSpec{
							Operation: types.VirtualDeviceConfigSpecOperationRemove,
							Device:    device,
						})
						log.Printf("[INFO] Removed USB controller: %s", oldUSBVersion)
						rebootRequired = true
					}
				}
			}
		}

		// Add new USB controller devices.
		for _, usbControllerInterface := range newUSB {
			usbController := usbControllerInterface.(map[string]interface{})
			usbVersion := usbController["version"].(string)

			var device types.BaseVirtualDevice

			switch usbVersion {
			case "2.0":
				if !usb2ControllerPresent {
					usb2ControllerPresent = true
					enabled := true // directly create a boolean variable
					device = &types.VirtualUSBController{
						VirtualController: types.VirtualController{
							VirtualDevice: types.VirtualDevice{
								Key: int32(keyCounter),
							},
						},
						EhciEnabled: &enabled, // assign address directly
					}
					rebootRequired = true
				}
			case "3.1", "3.2":
				if !usb3ControllerPresent {
					usb3ControllerPresent = true
					device = &types.VirtualUSBXHCIController{
						VirtualController: types.VirtualController{
							VirtualDevice: types.VirtualDevice{
								Key: int32(keyCounter),
							},
						},
					}
					rebootRequired = true
				}
			}

			if device != nil {
				spec.DeviceChange = append(spec.DeviceChange, &types.VirtualDeviceConfigSpec{
					Operation: types.VirtualDeviceConfigSpecOperationAdd,
					Device:    device,
				})
				log.Printf("[INFO] Added USB controller: %s", usbVersion)
				keyCounter--
			}
		}
		
		if rebootRequired {
			log.Printf("[INFO] USB controller changes require a virtual machine reboot.")
			_ = d.Set("reboot_required", true)
		}
	}
```